warehouse/postgres/postgres.go

```diff
@@ -0,0 +1,520 @@
+package postgres
+
+import (
+	"compress/gzip"
+	"database/sql"
+	"encoding/csv"
+	"errors"
+	"fmt"
+	"github.com/lib/pq"
+	"github.com/rudderlabs/rudder-server/config"
+	"github.com/rudderlabs/rudder-server/rruntime"
+	"github.com/rudderlabs/rudder-server/services/filemanager"
+	"github.com/rudderlabs/rudder-server/services/stats"
+	"github.com/rudderlabs/rudder-server/utils/logger"
+	"github.com/rudderlabs/rudder-server/utils/misc"
+	warehouseutils "github.com/rudderlabs/rudder-server/warehouse/utils"
+	uuid "github.com/satori/go.uuid"
+	"io"
+	"os"
+	"path/filepath"
+	"strings"
+	"sync"
+	"time"
+)
+
+var (
+	warehouseUploadsTable string
+	stagingTablePrefix    string
+	maxParallelLoads      int
+)
+
+const (
+	host     = "host"
+	dbName   = "database"
+	user     = "user"
+	password = "password"
+	port     = "port"
+	sslMode  = "sslMode"
+)
+
+var dataTypesMap = map[string]string{
+	"int":      "bigint",
+	"float":    "numeric",
+	"string":   "text",
+	"datetime": "timestamp with time zone",
+	"boolean":  "boolean",
+}
+
+type HandleT struct {
+	DbHandle      *sql.DB
+	Db            *sql.DB
+	Namespace     string
+	CurrentSchema map[string]map[string]string
+	Warehouse     warehouseutils.WarehouseT
+	Upload        warehouseutils.UploadT
+	CloudProvider string
+	ObjectStorage string
+}
+
+type credentialsT struct {
+	host       string
+	dbName     string
+	user       string
+	password   string
+	schemaName string
+	port       string
+	sslMode    string
+}
+
+type optionalCredsT struct {
+	schemaName string
+}
+
+var primaryKeyMap = map[string]string{
+	"users":                      "id",
+	"identifies":                 "id",
+	warehouseutils.DiscardsTable: "row_id",
+}
+var partitionKeyMap = map[string]string{
+	"users":                      "id",
+	"identifies":                 "id",
+	warehouseutils.DiscardsTable: "row_id, column_name, table_name",
+}
+
+func connect(cred credentialsT) (*sql.DB, error) {
+	url := fmt.Sprintf("user=%v password=%v host=%v port=%v dbname=%v sslmode=%v",
+		cred.user,
+		cred.password,
+		cred.host,
+		cred.port,
+		cred.dbName,
+		cred.sslMode)
+
+	var err error
+	var db *sql.DB
+	if db, err = sql.Open("postgres", url); err != nil {
+		return nil, fmt.Errorf("postgres connection error : (%v)", err)
+	}
+	return db, nil
+}
+
+func init() {
+	config.Initialize()
+	loadConfig()
+}
+func loadConfig() {
+	warehouseUploadsTable = config.GetString("Warehouse.uploadsTable", "wh_uploads")
+	stagingTablePrefix = "rudder_staging_"
+	maxParallelLoads = config.GetInt("Warehouse.postgres.maxParallelLoads", 1)
```

Comment:
change maxParallelLoads to bigger number and test for errors and mem usage/perf on wh master. Add in config.toml also. Also, update the default value in snowflake.go to 3 to match with config.toml